disktable.go

```diff
@@ -0,0 +1,650 @@
+
+					mu.Unlock()
+				}
+			}()
+		}
+
+		wg.Wait()
+
+		// Did we have any errors, if so return them.
+		select {
+		case err := <-errCh:
+			ch <- Result{Err: err}
+			return
+		default:
+		}
+
+		keysLookup := make([][]byte, 0, 100)
+
+		// Pull all the rows in the index that matched and send them back.
+		matches.Ascend(
+			func(item btree.Item) bool {
+				i := item.(btreeItem)
+				if int(i.count) == len(lookup) {
+					primaryKey := make([]byte, 8)
+					endian.PutUint64(primaryKey, i.primaryKey)
+					keysLookup = append(keysLookup, primaryKey)
+					// Ok, we have enough for a bulk lookup.
+					if len(keysLookup) == 1000 {
+						t.primary.db.View(t.retrievePrimaryRows(keysLookup, ch))
+						keysLookup = keysLookup[0:0]
+					}
+				}
+				return true
+			},
+		)
+
+		// Get any outstanding lookups that didn't have enough to do a bulk.
+		if len(keysLookup) > 0 {
+			t.primary.db.View(t.retrievePrimaryRows(keysLookup, ch))
+		}
+	}()
+
+	return ch, nil
+}
+
+func (t *Table) retrievePrimaryRows(keys [][]byte, ch chan Result) func(txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		for _, key := range keys {
+			item, err := txn.Get(key)
+			if err != nil {
+				return err
+			}
+			if !bytes.Equal(item.Key(), key) {
+				panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+			}
+			value, err := item.ValueCopy(nil)
+			if err != nil {
+				return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+			}
+			ch <- Result{Value: value}
+		}
+		return nil
+	}
+}
+
+type lookupResult struct {
+	i   uint64
+	err error
+}
+
+// indexLookup returns all matching keys in an index.
+func (t *Table) indexLookup(indexName string, key []byte) chan lookupResult {
+	index := t.indexByName[indexName]
+
+	realKey := finalKey(key, 0)
+
+	ch := make(chan lookupResult, 1)
+	if index.AllowDuplicates {
+		// multiMatch will close the channel.
+		go t.multiMatch(realKey, index, ch)
+	} else {
+		ch <- t.exactMatch(realKey, index)
+		close(ch)
+	}
+
+	return ch
+}
+
+// exactMatch does an exact match for a key in the index. key must have been finalKey()'d.
+func (t *Table) exactMatch(key []byte, index *Index) lookupResult {
+	var result lookupResult
+
+	f := func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(item.Key(), key) {
+			panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+		}
+		value, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+		}
+		result = lookupResult{i: endian.Uint64(value)}
+		return nil
+	}
+
+	if err := index.db.View(f); err != nil {
+		result.err = err
+	}
+	return result
+
+}
+
+// multiMatch matches all keys in an index that allows duplicates. key must have been finalKey()'d.
+func (t *Table) multiMatch(key []byte, index *Index, ch chan lookupResult) {
+	defer close(ch)
```

Comment:
if you do end up keeping this line, it should at least go into the docs that this function is gonna `close(ch)`